data_write/data_write.py

```diff
@@ -1184,34 +1184,55 @@ def main():
             final_integration = integration_meta.last_seqn_num
 
         if dsp_to_data_write in socks and socks[dsp_to_data_write] == zmq.POLLIN:
-            data = so.recv_bytes(dsp_to_data_write, options.dsp_to_dw_identity, printing)
-
-            if not first_time:
-                if data_parsing.sequence_num == final_integration:
-
-                    if integration_meta.experiment_name != current_experiment:
-                        data_write = DataWrite(options)
-                        current_experiment = integration_meta.experiment_name
-
-                    kwargs = dict(write_bfiq=args.enable_bfiq,
-                                           write_antenna_iq=args.enable_antenna_iq,
-                                           write_raw_rf=args.enable_raw_rf,
-                                           write_tx=args.enable_tx,
-                                           file_ext=args.file_type,
-                                           integration_meta=integration_meta,
-                                           data_parsing=data_parsing,
-                                           write_rawacf=args.enable_raw_acfs)
-                    thread = threading.Thread(target=data_write.output_data, kwargs=kwargs)
-                    thread.daemon = True
-                    thread.start()
-                    data_parsing = ParseData()
-
-            first_time = False
-
-            start = time.time()
-            data_parsing.update(data)
-            end = time.time()
-            printing("Time to parse: {} ms".format((end-start)*1000))
+            data = so.recv_bytes_from_any_iden(dsp_to_data_write)
+
+            processed_data = processeddata_pb2.ProcessedData()
+            processed_data.ParseFromString(data)
+
+            queued_sqns.append(processed_data)
+            # Check if any data processing finished out of order.
+            if processed_data.sequence_num != expected_sqn_num:
+                continue
+
+            sorted_q = sorted(queued_sqns, key=lambda x:x.sequence_num)
+
+            # This is needed to check that if we have a backlog, there are no more
+            # skipped sequence numbers we are still waiting for.
+            for i, pd in enumerate(sorted_q):
```

Comment:
This seems like it won't work... if I go through this loop on paper, expected_seq_num never goes beyond 0 for the case where I have sorted_q with seq numbers of (0,1,2,4,5). Maybe I'm missing something?

Reply:
Fixed